### PR TITLE
feat!: Increase Terraform and AWS provider minimum supported versions; update `auth` configuration schema

### LINKED ISSUE #10: cannot create proxy without `require_tls` = true
## Description

- [x] ✋ I have searched the open/closed issues and my issue is not listed.

## ⚠️ Note

Before you submit an issue, please perform the following first:

1. Remove the local `.terraform` directory (! ONLY if state is stored remotely, which hopefully you are following that best practice!): `rm -rf .terraform/`
2. Re-initialize the project root to pull down modules: `terraform init`
3. Re-attempt your terraform plan or apply and check if the issue still persists

## Versions

- Module version [Required]: 2.1.0
- Terraform version: 1.1.8
- Provider version(s): 4

## Reproduction Code [Required]

```
module "rds_proxy" {
  name = "cluster_name"
  source = "terraform-aws-modules/rds-proxy/aws"
  version = "2.1.0"

  # disable TLS check
  require_tls = false

  name                   = local.name
  iam_role_name          = local.name
  vpc_subnet_ids         = module.vpc.private_subnets
  vpc_security_group_ids = [module.rds_proxy_sg.security_group_id]

  db_proxy_endpoints = {
    read_write = {
      name                   = "read-write-endpoint"
      vpc_subnet_ids         = module.vpc.private_subnets
      vpc_security_group_ids = [module.rds_proxy_sg.security_group_id]
      tags                   = local.tags
    }
  }

  secrets = {
    "${local.db_username}" = {
      auth_scheme = "SECRETS"
      iam_auth    = "DISABLED"
      description = aws_secretsmanager_secret.superuser.description
      arn         = aws_secretsmanager_secret.superuser.arn
      kms_key_id  = aws_secretsmanager_secret.superuser.kms_key_id
    }
  }

  engine_family = "MYSQL"

  # Target Aurora cluster
  target_db_cluster     = true
  db_cluster_identifier = module.rds.cluster_id

  tags = local.tags
}
```

## Actual behavior

Error output

```
│ Error: Error creating DB Proxy: InvalidParameterValue: Must enable TLS, when IAM Auth is required
│ 	status code: 400, request id: de2093bc-e0b0-427c-9683-17d0bb843ece
```

### LINKED ISSUE #14: Documentation:  secrets.kms_key_id
The example shows this as a guid. It needs to be an arn, else the creation of the role will fail.

### LINKED ISSUE #16: Set IAM Auth per user not proxy wide
## Is your request related to a problem? Please describe.
Currently iam_auth is set proxy wide and pulled into the dynamic auth loop in the aws_db_proxy resource. With this approach you can't break down different users into IAM and non IAM authentication for the proxy.

## Describe the solution you'd like.
Make it so that iam_auth can be pulled from the secrets object into the dynamic auth block instead of being set globally for the whole proxy.

### LINKED ISSUE #18: Malformed IAM Policy
## Description

Small bug in module, an IAM Policy is referencing a secret id instead of the arn.
* **main.tf ::  line 138**
- [x] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

```
terraform {
  required_version = "=1.4.6"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.67.0"
    }
    null = {
      source  = "hashicorp/null"
      version = "3.2.1"
    }
    tls = {
      source  = "hashicorp/tls"
      version = "4.0.4"
    }
  }
}
```

## Reproduction Code

```
module "rds_proxy" {
  source                  = "terraform-aws-modules/rds-proxy/aws"
  version                 = "~>2.1.2"

  create_proxy            = true

  name                    = "${local.project_name}-rds-proxy"
  iam_role_name           = "${local.project_name}-rds-proxy-role"
  vpc_subnet_ids          = module.subnets.private_subnet_ids
  vpc_security_group_ids  = [module.rds_proxy_sg.security_group_id]

  db_proxy_endpoints      = {
    read_only = {
      name                   = "read-only-endpoint"
      vpc_subnet_ids         = module.subnets.private_subnet_ids
      vpc_security_group_ids = [module.rds_proxy_sg.security_group_id]
      target_role            = "READ_ONLY"
      tags                   = local.rds_proxy_tags
    }
  }

  secrets                 = {
    "rdxproxyadmin" = {
      description = aws_secretsmanager_secret.rds_proxy.description
      arn         = aws_secretsmanager_secret.rds_proxy.arn
      kms_key_id  = aws_secretsmanager_secret.rds_proxy.kms_key_id
    }
  }

  engine_family           = "POSTGRESQL"
  debug_logging           = true
  idle_client_timeout     = 300

  # Target Aurora cluster
  target_db_cluster       = true
  db_cluster_identifier   = module.aurora_postgresql_v2.cluster_id

  tags                    = local.rds_proxy_tags
}
```

* Try running module with IAM role

## Expected behavior
* IAM Policy to be created 

## Actual behavior

* Error: Malformed IAM Policy - Resource must be "*" or ARN

### LINKED ISSUE #19: Can't have more than 1 secret
Can't have more than 1 secret for the input variable **secrets**. 
It errors out with the following:
`│ Error: Insufficient auth blocks
│
│   on .terraform/modules/rds_proxy/main.tf line 13, in resource "aws_db_proxy" "this":
│   13: resource "aws_db_proxy" "this" {
│
│ At least 1 "auth" blocks are required.
╵
╷
│ Error: Invalid dynamic for_each value
│
│   on .terraform/modules/rds_proxy/main.tf line 26, in resource "aws_db_proxy" "this":
│   26:     for_each = var.secrets
│
│ Cannot use a map of object value in for_each. An iterable collection is required.`

Thanks in advance

### PR DESCRIPTION
**Pull request by jatalocks**
## Description

- Update Terraform minimum supported version to 1.0
- Update AWS provider minimum supported version to 5.0
- The manner in which authentication is configured has changed - previously auth settings were provided under `secrets` in conjunction with `auth_scheme` and `iam_auth` variables. Now, auth settings are provided under the `auth` variable for multiple auth entries.
- See the [UPGRADE-3.0.md](https://github.com/jatalocks/terraform-aws-rds-proxy/blob/master/UPGRADE-3.0.md) guide for further details

## Motivation and Context

- Allow users to better control the authentication settings of the proxy
- Resolves #10
- Resolves #14 
- Resolves #16 
- Resolves #18
- Resolves #19 

## Breaking Changes

- Yes; see [upgrade guide](https://github.com/jatalocks/terraform-aws-rds-proxy/blob/master/UPGRADE-3.0.md)

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by AlKapkone** [comment]
@antonbabenko Could you please take a look, because it's very important for me too. Thanks a lot, sorry to bother you.

**Comment by AlKapkone** [comment]
@amitai-devops @jatalocks Hi guys. 
Do you need help from me to update this MR?

**Comment by jatalocks** [comment]
> @amitai-devops @jatalocks Hi guys. 
> Do you need help from me to update this MR?

Hey, I'm the same person, just two accounts 😅, I can get to it when I can but I wouldn't say no if someone wants to commit to this PR

**Comment by AlKapkone** [comment]
@jatalocks Merge please, I hope this helps to merge these changes to the original module.
https://github.com/jatalocks/terraform-aws-rds-proxy/pull/1

File: versions.tf
Comment (antonbabenko) [inline versions.tf:2]: Please make the same changes in all `versions.tf` (including examples).

**Review by antonbabenko — CHANGES_REQUESTED** [review]

File: versions.tf
Comment (AlKapkone) [reply versions.tf:2]: Yes, sure, thx

Comment (AlKapkone) [reply versions.tf:2]: Fixed.

**Review by AlKapkone — APPROVED** [review]

File: versions.tf
Comment (AlKapkone) [reply versions.tf:2]: I will update VPC module version in examples to fix TF validate command.

**Comment by bryantbiggs** [comment]
@AlKapkone I'll update the examples since we've have a number of module changes - one sec

**Comment by bryantbiggs** [comment]
Validated on `mysql-iam-cluster` and `mysql-iam-instance` which should also cover the postgresql equivalents as well

@antonbabenko please take a look when you get a chance

File: UPGRADE-3.0.md
Comment (antonbabenko) [inline UPGRADE-3.0.md:67]: Missing closing `}` in `auth`

File: UPGRADE-3.0.md
Comment (antonbabenko) [inline UPGRADE-3.0.md:71]:
```suggestion
+    }
+  }
```

**Review by antonbabenko — APPROVED** [review]

File: UPGRADE-3.0.md
Comment (antonbabenko) [inline UPGRADE-3.0.md:74]:
```suggestion
```

**Review by antonbabenko — APPROVED** [review]

**Comment by AlKapkone** [comment]
@jatalocks Merge, please

**Comment by jatalocks** [comment]
> @jatalocks Merge, please

I am waiting for maintainer to merge, I'm not a maintainer

**Comment by AlKapkone** [comment]
@antonbabenko Do we need something else to add or we can merge it?

**Comment by antonbabenko** [comment]
This PR is included in version 3.0.0 :tada:

**Pull request merged**
